### ISSUE TITLE
xdrrec_create() type cast error: char* vs. void*

### ISSUE THREAD

**Issue by afaerber** [dyninst/dyninst#438]
I am trying to package dyninst 9.3.2 for openSUSE Tumbleweed and running into this build error:

```
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/string-regex.C:34:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h: In function 'void P_xdrrec_create(XDR*, u_int, u_int, caddr_t, xdr_rd_func, xdr_wr_func)':
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:45: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                            [   61.648111] serial8250: too much work for irq4
[   69s]   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 5 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/string-regex.C:34:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:80: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 6 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Ident.h:47,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Ident.C:36:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h: In function 'void P_xdrrec_create(XDR*, u_int, u_int, caddr_t, xdr_rd_func, xdr_wr_func)':
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:45: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 5 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Ident.h:47,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Ident.C:36:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:80: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 6 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/lprintf.C:40:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h: In function 'void P_xdrrec_create(XDR*, u_int, u_int, caddr_t, xdr_rd_func, xdr_wr_func)':
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:45: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 5 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/lprintf.C:40:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:80: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 6 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/pathName.C:32:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h: In function 'void P_xdrrec_create(XDR*, u_int, u_int, caddr_t, xdr_rd_func, xdr_wr_func)':
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:45: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 5 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/pathName.C:32:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:80: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/B[   61.752263] serial8250: too much work for irq4
[   69s] UILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 6 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Timer.h:43,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Timer.C:33:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h: In function 'void P_xdrrec_create(XDR*, u_int, u_int, caddr_t, xdr_rd_func, xdr_wr_func)':
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:45: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 5 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/headers.h:56:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Timer.h:43,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/Timer.C:33:
[   69s] /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/linuxHeaders.h:284:80: error: invalid conversion from 'int (*)(char*, char*, int)' to 'int (*)(void*, void*, int)' [-fpermissive]
[   69s]    xdrrec_create(x, send_sz, rec_sz, handle, (int(*)(char*, char*, int))read_r, (int(*)(char*, char*, int))write_f);}
[   69s]                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[   69s] In file included from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.cxx:30:0,
[   69s]                  from /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx:4:
[   69s] /usr/include/rpc/xdr.h:361:15: note:   initializing argument 6 of 'void xdrrec_create(XDR*, u_int, u_int, void*, int (*)(void*, void*, int), int (*)(void*, void*, int))'
[   69s]  extern void   xdrrec_create(XDR *, u_int, u_int, void *,
[   69s]                ^~~~~~~~~~~~~
[   69s] make[2]: *** [common/CMakeFiles/common.dir/build.make:112: common/CMakeFiles/common.dir/src/string-regex.C.o] Error 1
[   69s] make[2]: *** Waiting for unfinished jobs....
[   69s] make[2]: *** [common/CMakeFiles/common.dir/build.make:212: common/CMakeFiles/common.dir/src/lprintf.C.o] Error 1
[   69s] make[2]: *** [common/CMakeFiles/common.dir/build.make:137: common/CMakeFiles/common.dir/src/Timer.C.o] Error 1
[   69s] make[2]: *** [common/CMakeFiles/common.dir/build.make:87: common/CMakeFiles/common.dir/src/Ident.C.o] Error 1
[   69s] make[2]: *** [common/CMakeFiles/common.dir/build.make:237: common/CMakeFiles/common.dir/src/pathName.C.o] Error 1
[   69s] [  3%] Building CXX object common/CMakeFiles/common.dir/src/debugOstream.C.o
[   69s] cd /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common && /usr/bin/c++  -DBOOST_ALL_NO_LIB=1 -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DCOMMON_LIB -DUSE_PARSE_API -DWITHOUT_SYMLITE -DWITH_SYMTAB_API -Darch_64bit -Darch_x86_64 -Dbug_force_terminate_failure -Dbug_syscall_changepc_rewind -Dcap_32_64 -Dcap_async_events -Dcap_binary_rewriter -Dcap_dwarf -Dcap_dynamic_heap -Dcap_fixpoint_gen -Dcap_liveness -Dcap_mutatee_traps -Dcap_noaddr_gen -Dcap_ptrace -Dcap_registers -Dcap_stack_mods -Dcap_stripped_binaries -Dcap_thread_db -Dcap_threads -Dcap_tramp_liveness -Dcommon_EXPORTS -Dos_linux -Dx86_64_unknown_linux2_4 -I/usr/include/libdwarf -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/dataflowAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/dyninstAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/instructionAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/parseAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/patchAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/proccontrol/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/stackwalk/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/symtabAPI/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/dwarf/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/elf/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/symlite/h -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2 -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/external -I/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src  -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -DNDEBUG -std=c++11  -fvisibility=hidden -fvisibility-inlines-hidden -W -Wall -Wpointer-arith -Wcast-qual -Woverloaded-virtual -Wcast-align -Wno-non-template-friend -Wno-unused-local-typedefs -Wno-deprecated-declarations -O2 -g  -fPIC   -Winvalid-pch -include /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build/common/cotire/common_CXX_prefix.hxx -o CMakeFiles/common.dir/src/debugOstream.C.o -c /home/abuild/rpmbuild/BUILD/dyninst-9.3.2/common/src/debugOstream.C
[   69s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build'
[   69s] make[1]: *** [CMakeFiles/Makefile2:227: common/CMakeFiles/common.dir/all] Error 2
[   69s] make[1]: Leaving directory '/home/abuild/rpmbuild/BUILD/dyninst-9.3.2/build'
[   69s] make: *** [Makefile:133: all] Error 2
```
https://build.opensuse.org/package/live_build_log/home:a_faerber:hpc/dyninst/openSUSE_Tumbleweed/x86_64

**Comment by rockybulwinkle** [comment]
I'm seeing this issue as well when trying to build for Artix. Did you find a workaround?

**Comment by bwelton** [comment]
The change is likely simple (conversion to void * instead of char *). However the fix may break compilation on systems without this non-standard xdrrec_create function definition. I can give this a spin and see what happens.

**Comment by afaerber** [comment]
Maybe some preprocessor check can be used to support both? I think this was about code moving from glibc to libtirpc?

**Comment by bwelton** [comment]
Maybe, we generally like to avoid preprocessor checks if possible for code maintainability purposes. However, it may be required in this instance.

**Comment by mxz297** [comment]
Closed by #496
